File: data structures and algorithms/java/08_026.java
import java.util.*;

/**
 * Represents a cache operation with timing information
 */
class Operation {
   int arrivalTime;
   int targetNode;
   int sourceNode;
   String key;
   int value;
   int opTimestamp;
   
   Operation(int arrivalTime, int targetNode, int sourceNode, 
         String key, int value, int opTimestamp) {
      this.arrivalTime = arrivalTime;
      this.targetNode = targetNode;
      this.sourceNode = sourceNode;
      this.key = key;
      this.value = value;
      this.opTimestamp = opTimestamp;
   }
}

/**
 * Simulates eventual consistency in a distributed cache
 */
public class DistributedCacheConsistency {
   /**
    * Compute final cache state after operations propagate
    */
   public static List<Map<String, Integer>> finalCacheState(
         int n, int[][] operationsData, int[][] delays) {
      // Convert raw data to operation objects
      List<Operation> operations = new ArrayList<>();
      for (int[] op : operationsData) {
         int nodeId = op[0];
         String key = "key" + op[1]; // Convert to string key
         int value = op[2];
         int timestamp = op[3];
         operations.add(new Operation(0, nodeId, nodeId, key, 
               value, timestamp));
      }
      
      // Initialize node caches and timestamp tracking
      List<Map<String, Integer>> caches = new ArrayList<>();
      List<Map<String, Integer>> timestamps = new ArrayList<>();
      for (int i = 0; i < n; i++) {
         caches.add(new HashMap<>());
         timestamps.add(new HashMap<>());
      }
      
      // Event queue ordered by arrival time
      PriorityQueue<Operation> eventQueue = 
            new PriorityQueue<>(
               Comparator.comparingInt(op -> op.arrivalTime)
            );
      
      // Schedule initial operations and their propagation
      for (Operation op : operations) {
         int nodeId = op.targetNode;
         
         // Add local operation
         eventQueue.add(op);
         
         // Schedule propagation to other nodes with delays
         for (int targetNode = 0; targetNode < n; targetNode++) {
            if (targetNode != nodeId) {
               int arrivalTime = op.opTimestamp + 
                     delays[nodeId][targetNode];
               eventQueue.add(new Operation(
                     arrivalTime, 
                     targetNode, 
                     nodeId, 
                     op.key, 
                     op.value, 
                     op.opTimestamp));
            }
         }
      }
      
      // Process all operations in time order
      while (!eventQueue.isEmpty()) {
         Operation op = eventQueue.poll();
         int targetNode = op.targetNode;
         String key = op.key;
         int value = op.value;
         int opTimestamp = op.opTimestamp;
         
         // Apply operation if newer (last write wins)
         Integer currentTimestamp = 
               timestamps.get(targetNode).get(key);
         if (currentTimestamp == null || 
               currentTimestamp < opTimestamp) {
            caches.get(targetNode).put(key, value);
            timestamps.get(targetNode).put(key, opTimestamp);
         }
      }
      
      return caches;
   }
   
   /**
    * Test the distributed cache simulator
    */
   public static void main(String[] args) {
      int nodes = 3;
      int[][] operations = {
         {0, 0, 10, 1},  // Node 0 sets key0=10 at time 1
         {1, 1, 20, 2},  // Node 1 sets key1=20 at time 2
         {2, 0, 30, 3},  // Node 2 sets key0=30 at time 3
         {0, 1, 40, 4}   // Node 0 sets key1=40 at time 4
      };
      int[][] delays = {
         {0, 2, 3},  // Delays from node 0 to others
         {2, 0, 1},  // Delays from node 1 to others
         {3, 1, 0}   // Delays from node 2 to others
      };
      
      List<Map<String, Integer>> result = 
         finalCacheState(nodes, operations, delays);
      
      // Print final state of each node's cache
      for (int i = 0; i < result.size(); i++) {
         System.out.println("Node " + i + ": " + result.get(i));
      }
   }
}

File: data structures and algorithms/cpp/08_026.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <string>

struct Operation {
    int arrivalTime;
    int targetNode;
    int sourceNode;
    std::string key;
    int value;
    int opTimestamp;
    
    Operation(int arrivalTime, int targetNode, 
             int sourceNode, const std::string& key, 
             int value, int opTimestamp)
        : arrivalTime(arrivalTime), 
          targetNode(targetNode), 
          sourceNode(sourceNode),
          key(key), value(value), 
          opTimestamp(opTimestamp) {}
          
    bool operator>(const Operation& other) const {
        return arrivalTime > other.arrivalTime;
    }
};

std::vector<std::map<std::string, int>> finalCacheState(
        int n, 
        const std::vector
            std::tuple<int, std::string, int, int>>& operations, 
        const std::vector<std::vector<int>>& delays) {
    
    // Initialize caches for each node
    std::vector<std::map<std::string, int>> caches(n);
    std::vector<std::map<std::string, int>> timestamps(n);
    
    // Priority queue for events with timestamps
    std::priority_queue
        Operation, 
        std::vector<Operation>, 
        std::greater<Operation>> eventQueue;
    
    // Add initial operations to queue
    for (const auto& op : operations) {
        int nodeId = std::get<0>(op);
        std::string key = std::get<1>(op);
        int value = std::get<2>(op);
        int timestamp = std::get<3>(op);
        
        // Process operation at source node immediately
        eventQueue.emplace(timestamp, nodeId, nodeId, 
                          key, value, timestamp);
        
        // Schedule propagation to other nodes
        for (int targetNode = 0; targetNode < n; 
             targetNode++) {
            if (targetNode != nodeId) {
                int arrivalTime = timestamp + 
                    delays[nodeId][targetNode];
                eventQueue.emplace(arrivalTime, targetNode, 
                                  nodeId, key, value, 
                                  timestamp);
            }
        }
    }
    
    // Process all events
    while (!eventQueue.empty()) {
        Operation op = eventQueue.top();
        eventQueue.pop();
        
        int targetNode = op.targetNode;
        const std::string& key = op.key;
        int value = op.value;
        int opTimestamp = op.opTimestamp;
        
        // Apply if newer than what we have
        auto it = timestamps[targetNode].find(key);
        if (it == timestamps[targetNode].end() || 
            it->second < opTimestamp) {
            caches[targetNode][key] = value;
            timestamps[targetNode][key] = opTimestamp;
        }
    }
    
    return caches;
}

void printResult(
        const std::vector<std::map<std::string, int>>& 
        caches) {
    std::cout << "[";
    for (size_t i = 0; i < caches.size(); i++) {
        std::cout << "{";
        bool first = true;
        for (const auto& entry : caches[i]) {
            if (!first) std::cout << ", ";
            std::cout << "\"" << entry.first << "\": " 
                      << entry.second;
            first = false;
        }
        std::cout << "}";
        if (i < caches.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
}

int main() {
    int nodes = 3;
    std::vector<std::tuple<int, std::string, int, int>> 
        operations = {
            {0, "x", 10, 1},  // Node 0 sets x=10 at time 1
            {1, "y", 20, 2},  // Node 1 sets y=20 at time 2
            {2, "x", 30, 3},  // Node 2 sets x=30 at time 3
            {0, "y", 40, 4}   // Node 0 sets y=40 at time 4
        };
    std::vector<std::vector<int>> delays = {
        {0, 2, 3},  // Delays from node 0 to others
        {2, 0, 1},  // Delays from node 1 to others
        {3, 1, 0}   // Delays from node 2 to others
    };
    
    std::vector<std::map<std::string, int>> result = 
        finalCacheState(nodes, operations, delays);
    printResult(result);
    
    return 0;
}